### PR TITLE
Bindnode fixes continued

### PR DESCRIPTION
**Pull request by warpfork**
Like it says on the tin.

Details in commit messages.

### CODE REVIEW

**Comment by warpfork** [comment]
I've got one more fairly nasty discovery: our approach to unions is problematic *again*: because it is now required to use pointer fields, if one wants to use a union as a map key, one is going to have a bad day.  (As long as you hold on to literally the the values originally used as keys, you can use them to look things up again; but you'll have a hard time producing any other equivalent value, and that's probably going to result in massive frustration and nasty bugs, etc.)

We'll have to reintroduce the ability to use an integer field as the in-memory discriminator, and allow non-pointer fields for the members.  I think we can make this optional and just engage it if we find such a field.  (And then still probably document some warnings about this, because it'll be in the court of the person authoring the go types to not put themselves in a bad situation.) I'll probably look into doing this tomorrow.

File: node/bindnode/repr.go [hunk 1]
```diff
@@ -53,8 +53,8 @@ func (w *_nodeRepr) Kind() datamodel.Kind {
 		return datamodel.Kind_Map
 	case schema.UnionRepresentation_Kinded:
 		haveIdx, _ := unionMember(w.val)
-		mtyp := w.schemaType.(*schema.TypeUnion).Members()[haveIdx]
-		return mtyp.TypeKind().ActsLike()
+		mtyp := w.schemaType.(*schema.TypeUnion).Members()[haveIdx] // TODO this is fragile: panics with index-out-of-range if the user has nil'd all fields.
```
Comment (willscott) [inline node/bindnode/repr.go:56]: put a `, ok` and an error for out-of-range rather than panic?

File: node/bindnode/repr.go [hunk 2]
```diff
@@ -224,19 +226,57 @@ func (w *_nodeRepr) MapIterator() datamodel.MapIterator {
...
+			}
+		default:
+			panic(fmt.Sprintf("TODO: mapitr.repr for typekind %s", w.schemaType.TypeKind()))
+		}
 	default:
 		panic(fmt.Sprintf("TODO: %T", stg))
 	}
 }
 
+type _mapIteratorRepr _mapIterator
```
Comment (willscott) [inline node/bindnode/repr.go:244]: comment to explain why this wrapping and not just calling w's MapIterator is needed would be helpful

**Review by willscott — APPROVED** [review]
the additions here look reasonable

**Comment by warpfork** [comment]
I was too optimistic in saying I'd work on the union thing immediately.  I'm going to extract that to an issue instead.

File: node/bindnode/repr.go (hunk 1 above)
Comment (warpfork) [reply node/bindnode/repr.go:56]: Nontrivial.  This method can't return errors.  It's a core interface method, and a break-the-universe-badly one if we changed its return type.

**Comment by warpfork** [comment]
More tests for this would be desirable, but I'm going to hope we can backfill those in the future.  (And also, it should eventually be easier to do that if we get schema-schema wired up to be data-driven; then we can make purely data-driven language-agnostic tests, and gosh that would be nice.)  Meanwhile, as a stopgap, I promise I've at least tested all these myself in downstreams.

I'm also somewhat attached to these commits already, for the same reason (I've started depending on them by commit hash in a downstream already), so I'm going to boldly merge them, and we can continue improving in future PRs.

**Comment by mvdan** [comment]
All of these fixes make sense to me. I think we should start thinking about more table-driven tests before we continue doing more fixes, otherwise we're just building on quicksand.

**Pull request merged**